modApiServer/test/org/aion/api/server/ApiUtilTests.java

```diff
@@ -0,0 +1,160 @@
+        resultLength = 6;
+        vers = RandomUtils.nextBytes(1)[0];
+        retCode = RandomUtils.nextBytes(1)[0];
+        hash = RandomUtils.nextBytes(ApiUtil.HASH_LEN);
+        error = RandomUtils.nextBytes(errorLength);
+        result = RandomUtils.nextBytes(resultLength);
+    }
+
+    @Test
+    public void TestNulls() {
+        System.out.println("run TestNulls.");
+
+        assertNull(ApiUtil.toReturnHeader(vers, retCode, null));
+        assertNull(ApiUtil.toReturnHeader(vers, retCode, null, null));
+        assertNull(ApiUtil.toReturnHeader(vers, retCode, null, null, null));
+        assertNull(ApiUtil.combineRetMsg(null, null));
+        assertNull(ApiUtil.combineRetMsg(null, (byte) 1));
+        assertNull(ApiUtil.getApiMsgHash(null));
+    }
+
+    @Test
+    public void TestHeader() {
+        System.out.println("run TestHeader.");
+
+        byte[] header = ApiUtil.toReturnHeader(vers, retCode);
+        assertEquals(vers, header[0]);
+        assertEquals(retCode, header[1]);
+        assertEquals(0, header[2]);
+    }
+
+
+    @Test
+    public void TestHeaderHash() {
+        System.out.println("run TestHeaderHash.");
+
+        byte[] header = ApiUtil.toReturnHeader(vers, retCode, hash);
+        assertEquals(vers, header[0]);
+        assertEquals(retCode, header[1]);
+        assertEquals(1, header[2]);
+        assertArrayEquals(hash, Arrays.copyOfRange(header, 3, header.length));
+
+    }
+
+    @Test
+    public void TestHeaderHashError() {
+        System.out.println("run TestHeaderHashError.");
+
+        byte[] header = ApiUtil.toReturnHeader(vers, retCode, hash, error);
+        assertEquals(vers, header[0]);
+        assertEquals(retCode, header[1]);
+        assertEquals(1, header[2]);
+        assertArrayEquals(hash, Arrays.copyOfRange(header, 3, ApiUtil.HASH_LEN + 3));
+        assertEquals(error.length, header[ApiUtil.HASH_LEN + 3]);
+        assertArrayEquals(error, Arrays.copyOfRange(header, ApiUtil.HASH_LEN + 4, header.length));
+
+        error = new byte[0];
+        header = ApiUtil.toReturnHeader(vers, retCode, hash, error);
+        assertEquals(vers, header[0]);
+        assertEquals(retCode, header[1]);
+        assertEquals(1, header[2]);
+        assertArrayEquals(hash, Arrays.copyOfRange(header, 3, header.length - 1));
+        assertEquals(0, header[header.length - 1]);
+
+    }
+
+    @Test
+    public void TestHeaderHashErrorResult() {
+        System.out.println("run TestHeaderHashErrorResult.");
+
+        byte[] header = ApiUtil.toReturnHeader(vers, retCode, hash, error, result);
+        assertEquals(vers, header[0]);
+        assertEquals(retCode, header[1]);
+        assertEquals(1, header[2]);
+        assertArrayEquals(hash, Arrays.copyOfRange(header, 3, ApiUtil.HASH_LEN + 3));
+        assertEquals(error.length, header[ApiUtil.HASH_LEN + 3]);
+        assertArrayEquals(error, Arrays.copyOfRange(header, ApiUtil.HASH_LEN + 4, ApiUtil.HASH_LEN + errorLength + 4));
```

Comment:
Out of curiosity, how did you determine these constants?  i.e. + 4, + 3, etc

Reply:
I reverse-calculated them from the functions in ApiUtil.java, which let me figure out exactly what the contents of the headers are.